docs/limitations.rst

```diff
@@ -92,100 +93,161 @@ the "sync-barrier-sync" construct needs to be used.
-    Even with locking support, use of data sieving will drastically increase the
-    time and space overhead within UnifyFS, significantly decreasing application
-    performance. For ROMIO, set the hints ``romio_ds_write disable`` and
-    ``romio_ds_read disable`` to disable data sieving.
+    ROMIO uses ``fcntl()`` to support its data sieving optimization.
+    One must disable ROMIO data sieving when using UnifyFS.
+    To disable data sieving, one can set the following ROMIO hints:
+
+    - ``romio_ds_read disable``
+    - ``romio_ds_write disable``
 
-.. topic:: Shared File Pointers
+    These hints can be set in the ``MPI_Info`` object when opening a file,
+    e.g.,:
 
-    Avoid using shared file pointers in MPI-I/O under UnifyFS as they require
-    file locking to implement.
+    MPI_Info info;
+    MPI_Info_create(&info);
+    MPI_Info_set(info, "romio_ds_read",  "disable");
+    MPI_Info_set(info, "romio_ds_write", "disable");
+    MPI_File_open(comm, filename, amode, info, &fh);
+    MPI_Info_free(&info);
+
+    or the hints may be listed in a `ROMIOHINTS ROMIO_HINTS file`, e.g.,:
+
+    >>: cat romio_hints.txt
+    romio_ds_read disable
+    romio_ds_write disable
+
+    >>: export ROMIO_HINTS="romio_hints.txt"
+
+.. topic:: MPI-I/O Shared File Pointers
+
+    ROMIO uses file locking to support MPI-I/O shared file pointers.
+    One cannot use MPI-I/O shared file pointers when using UnifyFS.
     Functions that use shared file pointers include:
 
     - ``MPI_File_write_shared()``
     - ``MPI_File_read_shared()``
     - ``MPI_File_write_ordered()``
     - ``MPI_File_read_ordered()``
 
-File Locking Workarounds
-^^^^^^^^^^^^^^^^^^^^^^^^
+---------------------------
+HDF5 Limitations
+---------------------------
 
-UnifyFS doesn't provide any direct workarounds for anything that requires file
-locking. Simply disable atomicity and data sieving and avoid using shared file
-pointers to get around this.
+HDF5 uses MPI-I/O.
+In addition to restrictions that are specific to HDF5,
+one must follow any restrictions associated with the
+underlying MPI-I/O implementation.
+In particular, if the MPI library uses ROMIO for its MPI-I/O implementation,
+one should adhere to any limitations noted above for ROMIO.
 
-In the future, UnifyFS may add support for file locking. However, it is strongly
-suggested to avoid file locking unless the application cannot run properly
-without its use. Enabling file lock support within UnifyFS will result in
-decreased I/O performance for the application.
+Synchronization
+***************
+
+HDF5 FILE_SYNC
+""""""""""""""
+
+HDF5 provides a configuration option that internally calls ``MPI_File_sync()``
+after every collective HDF write operation as needed by MPI-I/O.
+Set the environment variable ``HDF5_DO_MPI_FILE_SYNC=1`` to enable this option.
+
+.. Note::
+
+    This option will soon be available in the `HDF5 develop branch`_ as well as
+    in the next HDF5 release.
```

Comment:
We should list an HDF5 version here if we can, since "next release" might have happened now.

Reply:
This appears to have been merged into the HDF5 `develop` branch in July of 2022 (https://github.com/HDFGroup/hdf5/pull/1801).
I'm seeing it in the three releases since then (v1.13.2, v1.13.3, and v1.14.0).

Reply:
I mentioned this in my review below as well, but is this the config HDF5 created to do the same thing as the `romio_visibility_immediate` hint (once available)?
That hint appears to now be complete and will be available in MPICH v4.1

Reply:
> I mentioned this in my review below ([#756 (comment)](https://github.com/LLNL/UnifyFS/pull/756#discussion_r1068801867)) as well, but is this the config HDF5 created to do the same thing as the `romio_visibility_immediate` hint (once available)?
> That hint appears to now be complete and will be available in MPICH v4.1

I need to double check since it's been a while, but I think this is a read-only hint that informs the MPI-I/O user (HDF5) as to whether they need to execute a sync-barrier-sync operation in order to make data visible.  I think this hint returns ``true`` on Lustre and GPFS, meaning that data is immediately visible and a sync-barrier-sync is not required, whereas it should return ``false`` on UnifyFS indicating that a sync-barrier-sync is necessary.